scipy/spatial/transform/_rotation_cy.pyx

```diff
@@ -0,0 +1,1425 @@
+
+    return np.asarray(matrix)
+
+
+@cython.embedsignature(True)
+@cython.boundscheck(False)
+@cython.wraparound(False)
+def as_rotvec(double[:, :] quat, bint degrees=False) -> double[:, :]:
+    cdef Py_ssize_t num_rotations = len(quat)
+    cdef double angle, scale, angle2
+    cdef double[:, :] rotvec = _empty2(num_rotations, 3)
+    cdef double[:] quat_single
+
+    for ind in range(num_rotations):
+        quat_single = quat[ind, :].copy()
+        _quat_canonical_single(quat_single)  # w > 0 ensures that 0 <= angle <= pi
+
+        angle = 2 * atan2(_norm3(quat_single), quat_single[3])
+
+        if angle <= 1e-3:  # small angle Taylor series expansion
+            angle2 = angle * angle
+            scale = 2 + angle2 / 12 + 7 * angle2 * angle2 / 2880
+        else:  # large angle
+            scale = angle / sin(angle / 2)
+
+        rotvec[ind, 0] = scale * quat_single[0]
+        rotvec[ind, 1] = scale * quat_single[1]
+        rotvec[ind, 2] = scale * quat_single[2]
+
+    if degrees:
+        rotvec = np.rad2deg(rotvec)
+
+    return np.asarray(rotvec)
+
+
+@cython.embedsignature(True)
+@cython.boundscheck(False)
+@cython.wraparound(False)
+def as_mrp(double[:, :] quat) -> double[:, :]:
+    cdef Py_ssize_t num_rotations = len(quat)
+    cdef double[:, :] mrps = _empty2(num_rotations, 3)
+    cdef int sign
+    cdef double denominator
+
+    for ind in range(num_rotations):
+
+        # Ensure we are calculating the set of MRPs that correspond
+        # to a rotation of <= 180
+        sign = -1 if quat[ind, 3] < 0 else 1
+
+        denominator = 1 + sign * quat[ind, 3]
+        for i in range(3):
+            mrps[ind, i] = sign * quat[ind, i] / denominator
+
+    return np.asarray(mrps)
+
+
+@cython.embedsignature(True)
+@cython.boundscheck(False)
+@cython.wraparound(False)
+def as_euler(double[:, :] quat, seq, bint degrees=False) -> double[:, :]:
+    # Prepare axis sequence to call Euler angles conversion algorithm.
+    if len(seq) != 3:
+        raise ValueError("Expected 3 axes, got {}.".format(seq))
+
+    intrinsic = (re.match(r'^[XYZ]{1,3}$', seq) is not None)
+    extrinsic = (re.match(r'^[xyz]{1,3}$', seq) is not None)
+    if not (intrinsic or extrinsic):
+        raise ValueError("Expected axes from `seq` to be from "
+                            "['x', 'y', 'z'] or ['X', 'Y', 'Z'], "
+                            "got {}".format(seq))
+
+    if any(seq[i] == seq[i+1] for i in range(2)):
+        raise ValueError("Expected consecutive axes to be different, "
+                            "got {}".format(seq))
+
+    seq = seq.lower()
+    if not extrinsic:
+        seq = seq[::-1]
+
+    # The algorithm assumes extrinsic frame transformations. The algorithm
+    # in the paper is formulated for rotation quaternions, which are stored
+    # directly by Rotation.
+    # Adapt the algorithm for our case by reversing both axis sequence and
+    # angles for intrinsic rotations when needed
```

Comment:
Marking to myself that this algorithm has changed, need to come back and review more thoroughly.

Reply:
Ok, I see this is just the switch to _compute_euler_from_quat. Note that "the paper" is no longer in any comments, so should remove that bit of this comment.

Reply:
Or even better, update to cite the paper: https://journals.plos.org/plosone/article?id=10.1371/journal.pone.0276302